From 72a3791091b8f18d3e248b7127f2b81822c46139 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E7=99=BD=E6=B3=BD?= <patrickjiang0530@gmail.com>
Date: Thu, 20 Jan 2022 00:44:14 +0800
Subject: fix: when service entry's host has tailing period, smart dns can not
 resolve it (#36891)

* determine if the hostname ends with a period

* format code
---
 pkg/dns/client/dns.go      | 5 ++++-
 pkg/dns/client/dns_test.go | 9 +++++++++
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/pkg/dns/client/dns.go b/pkg/dns/client/dns.go
index 553c59ca08..5a5873b129 100644
--- a/pkg/dns/client/dns.go
+++ b/pkg/dns/client/dns.go
@@ -196,7 +196,10 @@ func (h *LocalDNSServer) UpdateLookupTable(nt *dnsProto.NameTable) {
 		if ni.Registry == string(provider.Kubernetes) {
 			altHosts = generateAltHosts(hostname, ni, h.proxyNamespace, h.proxyDomain, h.proxyDomainParts)
 		} else {
-			altHosts = map[string]struct{}{hostname + ".": {}}
+			if !strings.HasSuffix(hostname, ".") {
+				hostname += "."
+			}
+			altHosts = map[string]struct{}{hostname: {}}
 		}
 		ipv4, ipv6 := separateIPtypes(ni.Ips)
 		if len(ipv6) == 0 && len(ipv4) == 0 {
diff --git a/pkg/dns/client/dns_test.go b/pkg/dns/client/dns_test.go
index 66454f7246..16e381b4b3 100644
--- a/pkg/dns/client/dns_test.go
+++ b/pkg/dns/client/dns_test.go
@@ -217,6 +217,11 @@ func TestDNS(t *testing.T) {
 			expectResolutionFailure: dns.RcodeSuccess,
 			expected:                giantResponse[:29],
 		},
+		{
+			name:     "success: hostname with a period",
+			host:     "example.localhost.",
+			expected: a("example.localhost.", []net.IP{net.ParseIP("3.3.3.3").To4()}),
+		},
 	}
 
 	clients := []dns.Client{
@@ -514,6 +519,10 @@ func initDNS(t test.Failer) *LocalDNSServer {
 				Ips:      []string{"10.10.10.10"},
 				Registry: "External",
 			},
+			"example.localhost.": {
+				Ips:      []string{"3.3.3.3"},
+				Registry: "External",
+			},
 		},
 	})
 	t.Cleanup(testAgentDNS.Close)
-- 
2.35.3

